src/net/tcp/nbd.c

```diff
@@ -0,0 +1,1095 @@
+			return -ENOENT;
+
+		default:
+			DBGC ( nbd, "NBD %p option info failed (2),"
+			       " type = 0x%08x\n", nbd, opt_type );
+			return -EPROTO;
+		}
+	}
+
+	switch ( opt_type ) {
+	case NBD_REP_ACK:
+		if ( nbd->export_size == 0 || opt_len != 0 ) {
+			DBGC ( nbd, "NBD %p option info failed (3)\n", nbd );
+			return -EPROTO;
+		}
+
+		nbd_neg_done ( nbd );
+		break;
+
+	case NBD_REP_INFO:
+		if ( opt_len <= sizeof ( nbd-> rx_rep_info_export ) ) {
+			nbd->reply_info_length = opt_len;
+			nbd->rx_state = NBD_RX_NEG_REP_INFO;
+			break;
+		}
+		/* Fall through */
+	default:
+		DBGC2 ( nbd, "NBD %p option info ignored, type = 0x%08x\n",
+			nbd, opt_type );
+		nbd->discard_len = opt_len;
+		break;
+	}
+
+	return 0;
+}
+
+/**
+ * Process NBD_REP_INFO
+ *
+ * @v nbd		NBD session
+ * @v data		Received data
+ * @v len		Length of received data
+ * @ret rc		Return status code
+ */
+static int nbd_rx_neg_rep_info ( struct nbd_session *nbd ) {
+	DBGC2 ( nbd, "NBD %p rep info:\n", nbd );
+        DBGC2_HD ( nbd, &nbd->rx_rep_info_export, nbd->reply_info_length );
+
+	if ( ntohs ( nbd->rx_rep_info_export.type ) == NBD_INFO_EXPORT ) {
+		nbd->export_size =
+			ntohll ( nbd->rx_rep_info_export.export_size );
+		nbd->trans_flags =
+			ntohs ( nbd->rx_rep_info_export.trans_flags );
+	}
+
+	nbd->rx_state = NBD_RX_NEG_OPT_INFO;
+	return 0;
+}
+
+/**
+ * Process reply to NBD_CMD_?
+ *
+ * @v nbd		NBD session
+ * @v data		Received data
+ * @v len		Length of received data
+ * @ret rc		Return status code
+ */
+static int nbd_rx_trans_rep_cmd ( struct nbd_session *nbd ) {
+	struct nbd_command *nbdcmd = nbd->command;
+	uint32_t errno;
+	DBGCIO ( nbd, "NBD %p cmd %p reply:\n", nbd, nbdcmd );
+        DBGCIO_HD ( nbd, &nbd->rx_trans_reply, sizeof ( nbd->rx_trans_reply ) );
+
+	if ( nbdcmd == NULL ) {
+		DBGC ( nbd, "NBD %p no cmd to process\n", nbd );
+		return -EPROTO;
+	}
+
+	if ( ntohl ( nbd->rx_trans_reply.reply_magic ) != NBD_REPLY_MAGIC  ||
+	     nbd->rx_trans_reply.handle != (uint64_t) (intptr_t) nbdcmd ) {
+		DBGC ( nbd, "NBD %p cmd reply invalid\n", nbd );
+		return -EPROTO;
+	}
+
+	errno = ntohl ( nbd->rx_trans_reply.error );
+	if ( errno != 0 ) {
+		DBGC ( nbd, "NBD %p cmd errno: %d\n", nbd, errno );
+		return -EIO;
+	}
+
+	if ( nbdcmd->type == NBD_BLOCK_READ ) {
+		nbd->rx_state = NBD_RX_TRANS_DATA;
+	} else {
+		DBGCP ( nbd, "NBD %p cmd complete\n", nbd );
+		nbdcmd_close ( nbdcmd, 0 );
+	}
+
+	return 0;
+}
+
+/**
+ * Process data of NBD_CMD_READ
+ *
+ * @v nbd		NBD session
+ * @v iobuf		I/O buffer
+ * @ret rc		Return status code
+ */
+static int nbd_rx_trans_data ( struct nbd_session *nbd,
+			       struct io_buffer *iobuf ) {
+	struct nbd_command *nbdcmd = nbd->command;
+	size_t len = iob_len ( iobuf );
+	DBGCP ( nbd, "NBD %p cmd %p data len: %zd\n", nbd, nbdcmd, len );
+
+	if ( nbdcmd == NULL ) {
+		DBGC ( nbd, "NBD %p no cmd to process\n", nbd );
+		return -EPROTO;
+	}
+
+	/* Sanity checks */
+	if ( nbdcmd->data_offset + len > nbdcmd->data_len ) {
```

Comment:
Is this safe against integer overflow?

Reply:
Yes, because nbdcmd->data_len is limited by capacity.max_count

Reply:
This limit is imposed in sandev_rw function (sanboot.c). Conversely, using -1U as max_count can be dangerous.